pyro/infer/mcmc/nuts.py

```diff
@@ -0,0 +1,211 @@
+
+    References
+
+    [1] `The No-U-turn sampler: adaptively setting path lengths in Hamiltonian Monte Carlo`,
+    Matthew D. Hoffman, and Andrew Gelman
+
+    :param model: python callable containing pyro primitives.
+    :param float step_size: determines the size of a single step taken by the
+        verlet integrator while computing the trajectory using Hamiltonian
+        dynamics.
+    :param int max_tree_depth: maximum depth of binary tree during No-U-turn sampling.
+
+    Example::
+
+        true_coefs = Variable(torch.arange(1, 4))
+        data = Variable(torch.randn(2000, 3))
+        labels = dist.Bernoulli(logits=(true_coefs * data).sum(-1)).sample()
+
+        def model(data):
+            coefs_mean = Variable(torch.zeros(dim), requires_grad=True)
+            coefs = pyro.sample('beta', dist.Normal(coefs_mean, Variable(torch.ones(3))))
+            y = pyro.sample('y', dist.Bernoulli(logits=(coefs * data).sum(-1)), obs=labels)
+            return y
+
+        nuts_kernel = NUTS(model, step_size=0.0855)
+        mcmc_run = MCMC(nuts_kernel, num_samples=500, warmup_steps=100)
+        posterior = []
+        for trace, _ in mcmc_run._traces(data):
+            posterior.append(trace.nodes['beta']['value'])
+    """
+
+    def __init__(self, model, step_size=1, max_tree_depth=10):
+        # The default values for step_size and max_tree_depth are selected as in Stan
+        #     https://github.com/stan-dev/pystan/blob/develop/pystan/misc.py.
+        super(NUTS, self).__init__(model, step_size, num_steps=None)
+        self.max_tree_depth = max_tree_depth
+        self._reset()
+
+        # There are two conditions to stop doubling process:
+        #     + The trajectory is making a U-turn.
+        #     + The probability of the states becoming negligible: p(z, r) << u,
+        # here u is the "slice" variable introduced at the `self.sample(...)` method.
+        # Denote E_p = -log p(z, r), E_u = -log u, the second condition is equivalent to
+        #     dE := E_p - E_u >= some constant =: dE_max.
+        # This also suggests the notion "diverging" in the implemenation:
+        #     when the energy E_p diverges from E_u too much, we stop doubling.
+        # Here, as suggested in [1], we set dE_max = 1000.
+        self._dE_max = 1000
+
+    def _is_turning(self, z_left, r_left, z_right, r_right):
+        nodes = sorted(z_left)
+        z_left = torch.stack([z_left[name] for name in nodes])
+        r_left = torch.stack([r_left[name] for name in nodes])
+        z_right = torch.stack([z_right[name] for name in nodes])
+        r_right = torch.stack([r_right[name] for name in nodes])
+        dz = z_right - z_left
+        return (torch_data_sum(dz * r_left) < 0) or (torch_data_sum(dz * r_right) < 0)
+
+    def _build_basetree(self, z, r, log_slice, direction):
+        step_size = self.step_size if direction == 1 else -self.step_size
+        z_new, r_new = velocity_verlet(z, r, self._potential_energy, step_size)
```

Comment:
It does seem that for the same number of steps, we will be doing double the number of gradient (and model log prob eval) computations in `velocity_verlet` for NUTS. The for loop in the integrator reuses the computation in the previous step which we cannot do with `num_steps=1` and end up repeating the computation. Correct me if I am mistaken. I don't immediately see a clean way of caching this at the moment, but will think about it so that we can avoid this double cost.

Reply:
Whoa, a crucial point! Thanks for pointing it.